rpc/v2/json.go

```diff
@@ -0,0 +1,336 @@
+// Copyright 2015 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package v2
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/logger"
+	"github.com/ethereum/go-ethereum/logger/glog"
+)
+
+const (
+	jsonRPCVersion         = "2.0"
+	serviceMethodSeparator = "_"
+	subscribeMethod        = "eth_subscribe"
+	unsubscribeMethod      = "eth_unsubscribe"
+	notificationMethod     = "eth_subscription"
```

Comment:
This is ugly. 

Sorry for not seeing it earlier in the design phase, but funneling pub/sub through a single
magic method feels weird. My initial thought while looking at the API code was that subscribing
would just be a method call to e.g. `eth_newBlocks`. It doesn't work this way of course, but maybe
it should. 

I can see why the magic methods are required for your pub/sub scheme (there is no good way to handle unsubscribe inside the API handler because subscription management is implemented in package v2). One way to do it would be to define subscriptions as a channel+ID and moving subscription management to the API instead. That way, handlers would need to implement subscribe/unsubscribe themselves, but could even use an arbitrary event mechanism to implement them (currently use of package event is required). The role of `Server` would be reduced to converting events on the channel to notifications until it is closed.

```
package v2

type Subscription struct {
    ID []byte // assigned by API handle
    Method string // method name used for event notifications
    Events <-chan interface{} // closed when unsubscribed
}
```